File: 백준/Silver/22342. 계산 로봇/계산 로봇.cc
#include <bits/stdc++.h>
using namespace std;
#define ll long long
#define MOD 1000000007
#define MAX 100000
#define INF 987654321
int MAP[2000][2000];
int dp[2000][2000];
int main()
{
    ios_base::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);

    int n, m;
    cin >> n >> m;

    for (int i = 0; i < n; i++)
    {
        string s;
        cin >> s;
        for (int j = 0; j < m; j++)
            MAP[i][j] = s[j] - '0';
    }

    for (int i = 0; i < n; i++)
        dp[i][0] = MAP[i][0];

    for (int i = 1; i < m; i++)
    {
        for (int j = 0; j < n; j++)
        {

            if (i == m - 1)
                dp[j][i] = dp[j][i - 1];
            else
            {
                int temp = 0;
                for (int k = max(j - 1, 0); k <= min(j + 1, n - 1); k++)
                    temp = max(temp, dp[k][i - 1]);
                dp[j][i] = temp + MAP[j][i];
            }
        }
    }

    int answer = 0;

    for (int i = 0; i < n; i++)
    {
        answer = max(answer, dp[i][m - 1]);
    }

    cout << answer;
    return 0;
}

File: 백준/Silver/22342. 계산 로봇/계산 로봇.py
n, m = map(int, input().split())
MAP = [list(map(int, input().rstrip())) for _ in range(n)]
dp = [[0] * m for _ in range(n)]

for i in range(n):
    dp[i][0] = MAP[i][0]

for i in range(1, m - 1):
    for j in range(n):
        up = max(j - 1, 0)
        down = min(j + 1, n - 1)

        temp = 0
        for k in range(up, down + 1):
            temp = max(temp, dp[k][i - 1])

        dp[j][i] = temp + MAP[j][i]

answer = 0

for i in range(n):
    answer = max(answer, dp[i][m - 2])

print(answer)
